internal/pkg/secureboot/uki/internal/pe/native.go

```diff
@@ -0,0 +1,275 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package pe
+
+import (
+	"bytes"
+	"debug/pe"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+	"time"
+
+	"github.com/siderolabs/gen/xslices"
+
+	"github.com/siderolabs/talos/internal/pkg/secureboot"
+	"github.com/siderolabs/talos/pkg/imager/utils"
+)
+
+// AssembleNative is a helper function to assemble the PE file without external programs.
+//
+//nolint:gocyclo,cyclop
+func AssembleNative(srcPath, dstPath string, sections []Section) error {
+	in, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+
+	peFile, err := pe.NewFile(in)
+	if err != nil {
+		return err
+	}
+
+	defer in.Close() //nolint: errcheck
+
+	if peFile.COFFSymbols != nil {
+		return errors.New("COFF symbols are not supported")
+	}
+
+	if peFile.StringTable != nil {
+		return errors.New("COFF string table is not supported")
+	}
+
+	if peFile.Symbols != nil {
+		return errors.New("symbols are not supported")
+	}
+
+	_, err = in.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(dstPath)
+	if err != nil {
+		return fmt.Errorf("failed to create output: %w", err)
+	}
+
+	defer out.Close() //nolint: errcheck
+
+	// 1. DOS header
+	var dosheader [0x40]byte
+
+	_, err = in.ReadAt(dosheader[:], 0)
+	if err != nil {
+		return fmt.Errorf("failed to read DOS header: %w", err)
+	}
+
+	binary.LittleEndian.PutUint32(dosheader[0x3c:], 0x80)
+
+	_, err = out.Write(append(append(dosheader[:], bytes.Repeat([]byte{0}, 0x40)...), []byte("PE\x00\x00")...))
+	if err != nil {
+		return fmt.Errorf("failed to write DOS header: %w", err)
+	}
+
+	// 2. PE header and optional header
+	newFileHeader := peFile.FileHeader
+
+	timestamp, ok, err := utils.SourceDateEpoch()
+	if err != nil {
+		return fmt.Errorf("failed to get SOURCE_DATE_EPOCH: %w", err)
+	}
+
+	if !ok {
+		timestamp = time.Now().Unix()
+	}
+
+	newFileHeader.TimeDateStamp = uint32(timestamp)
+
+	// find the first VMA address
+	lastSection := peFile.Sections[len(peFile.Sections)-1]
+
+	header, ok := peFile.OptionalHeader.(*pe.OptionalHeader64)
+	if !ok {
+		return errors.New("failed to get optional header")
+	}
+
+	sectionAlignment := uint64(header.SectionAlignment - 1)
+	fileAlignment := uint64(header.FileAlignment - 1)
+
+	baseVirtualAddress := uint64(lastSection.VirtualAddress) + uint64(lastSection.VirtualSize)
+	baseVirtualAddress = (baseVirtualAddress + sectionAlignment) &^ sectionAlignment
+
+	newHeader := *header
+	newHeader.MajorLinkerVersion = 2
```

Comment:
should we just set with whatever objcopy does or set as `0` like ukify does?

Reply:
fixed this up, ignoring mismatch in the tests